### PR TITLE
Allow build numbers to be arbitrary strings or null

### PR DESCRIPTION
**Pull request by marcelstoer**
This required changing the `buildNumber` data type from `long` to `String`. @CDPrete this fixes your issue #75. Hence, your feedback is appreciated.

### CODE REVIEW

**Comment by marcelstoer** [comment]
@danielflower I have more new features coming that are based in this branch. Hence, I would appreciated if you could provide guiding feedback soon, thanks.

**Comment by marcelstoer** [comment]
@danielflower Besides this and #94 I have more improvements in mind for this plugin. Yet, I must say not even getting any feedback on PRs is very de-motivating. What are your plans with this project if I may ask?

**Comment by danielflower** [comment]
Hi Marcel, sorry I know it's frustrating to not get feedback. I just really don't have much time to go through things very often. Merging a pull request is a commitment on my behalf as I have to understand what is happening on code I have touched for years and then support those changes going forward, so I need to find enough time to go through changes. (The smaller the change, the easier to do.)

As for this change, this feels like quite a bit of a change from the current logic of "the build number gets incremented". I don't really understand what happens when it tries to increment a string.

**Comment by marcelstoer** [comment]
I'll happy to answer any questions - just keep'em coming.

> As for this change, this feels like quite a bit of a change from the current logic of "the build number gets incremented".

Most of changes are caused by the long-to-String switch for the build number (many method signatures, lots of tests). And then there are the new tests of course.

> I don't really understand what happens when it tries to increment a string.

To me the only really interesting corner case is if you switch from numeric build numbers (past) to alpha-numeric (now and possibly for the future) when the business version remains as-is.

`findTagWithHighestBuildNumber` is interesting here. See https://github.com/danielflower/multi-module-maven-release-plugin/pull/90/files#diff-98bd0c93f45924a297818a9f54e14d63R200 The "highest build _number_" is really just that: the largest non-alpha-numeric build number.

Or `nextBuildNumber` at https://github.com/danielflower/multi-module-maven-release-plugin/pull/90/files#diff-55ca5a18708be7ad8ecb7bceb23cb697R61. Skips over any non-numeric previous build numbers. Keep in mind that this method will only be called if you don't set a build number yourself!

File: pom.xml [hunk 1]
```diff
@@ -123,6 +123,15 @@
             <artifactId>maven-invoker</artifactId>
             <version>3.0.1</version>
         </dependency>
+        <dependency>
+            <!--
+            The maven-dependency-plugin wants this to be listed explicitly as it comes as a transitive dependency
+            of maven-model.
```
Comment (danielflower) [inline pom.xml]: So that means another thing is using this dependency already? I'd rather not risk the version dependency clashes. Seems like you just need an empty string, a null check, and a long check to replace this dependency.

File: src/main/java/com/github/danielflower/mavenplugins/release/Reactor.java [hunk 2]
```diff
@@ -190,20 +191,25 @@ static AnnotatedTag hasChangedSinceLastRelease(List<AnnotatedTag> previousTagsFo
...
-    private static AnnotatedTag tagWithHighestBuildNumber(List<AnnotatedTag> tags) {
-        AnnotatedTag cur = null;
+    // Ignores any tags with non-numeric build numbers and finds the one with the largest build number among the others.
+    // The tags are guaranteed to be for the same /business version/ i.e. it's likely a small list.
+    private static AnnotatedTag findTagWithHighestBuildNumber(List<AnnotatedTag> tags) {
+        AnnotatedTag tagWithHighestBuildNumber = null;
         for (AnnotatedTag tag : tags) {
-            if (cur == null || tag.buildNumber() > cur.buildNumber()) {
-                cur = tag;
+            if(StringUtils.isNumeric(tag.buildNumber())) {
```
Comment (danielflower) [inline src/main/java/com/github/danielflower/mavenplugins/release/Reactor.java]: As per comment in pom, I'd rather avoid dependencies if possible. The javadoc for this also says that this returning true doesn't guarantee Long.parseLong will work (e.g. if it's a really big number). It's fine to just not check and just try/catch the parseLong. There are no performance concerns etc here.

File: pom.xml (hunk 1 above)
Comment (marcelstoer) [reply pom.xml]: As this plugin relies on `maven-model` (they all do) all plugin code has access to `commons-lang3`.

Comment (danielflower) [reply pom.xml]: ...except now we have a dependency on the specific version of that dependency. That may be the same as maven-model now, but maybe not later.

Comment (marcelstoer) [reply pom.xml]: I'll address this.

Comment (marcelstoer) [reply pom.xml]: Removed the _use_ of the dependency.

File: src/main/java/com/github/danielflower/mavenplugins/release/Reactor.java (hunk 2 above)
Comment (marcelstoer) [reply src/main/java/com/github/danielflower/mavenplugins/release/Reactor.java]: Removed the _use_ of the dependency.

**Comment by marcelstoer** [comment]
All review comments addressed from my side.

**Comment by CDPrete** [comment]
> This required changing the `buildNumber` data type from `long` to `String`. @CDPrete this fixes your issue #75. Hence, your feedback is appreciated.

It looks promising to me.

**Comment by danielflower** [comment]
The reason I've not merged this so far is that in my mind this is a fundamental change to the plugin. In my mind, the plugin generates build numbers for you by adding 1 to the previous number. I'm probably missing something though.

Maybe some examples of build numbers in the pom.xml vs build numbers in a release?

e.g. currently if the pom version is `1.0-SNAPSHOT` then the first build is `1.0.0` and the second is `1.0.1`. This is what I can remember about this plugin. So what scenarios does this pull request cover?

**Comment by marcelstoer** [comment]
> in my mind this is a fundamental change to the plugin

Don't worry, it's not 😄

The original behavior does not change at all. Unless you (as a user) explicitly want so the plugin behavior remains as-is. There are situations, though, where this strict build number increment is undesirable. You could want your build number to be "A" or "foo" - or nothing at all. The Javadoc for [`buildNumber()` in `BaseMojo.java`](https://github.com/danielflower/multi-module-maven-release-plugin/pull/90/files#diff-d175bc6023cfe99909959499d43459ccR51) has a long description for the behavior. It goes without say that if you chose "A" in the _previous_ release the plugin won't be able to automatically increment that for the _current_ release.

Hence, to allow such flexibility I had to change the data type from `long` to `String`. That's the cause for many of the changes you see in this PR.

**Comment by danielflower** [comment]
Hi, I don't know if you are still interested in this, so long afterwards, but if you want to get it merged I think all it's missing is an explanation of when this can be used, why it would be used, and what happens to build numbers after using it (this lack of understanding on my behalf is why I took so long to take the time to understand this, so it's probably confusing to others too).

I think a section on https://danielflower.github.io/multi-module-maven-release-plugin/usage.html would be good.

Something like this?

> ## Specifying custom build numbers
>
> If you want to make a release that has version number with a string in it because ________
> To do this, set the parameter like so: `example`

Perhaps work in the clear explanation that @CDPrete originally wrote on #75 

> * if the buildNumber is specified:
>     *   if the string is not null and not empty, then attach the delimiter
>     *  try to parse it as Long and, if it works, just use parsed-value + 1
>     *   if the parsing fails, use the specified string
> *   else if no buildNumber is specified:
>      *   just use the business version alone

(Does the above logic hold for this PR?)

The documentation needs some concrete examples, e.g. if I have `1.0-SNAPSHOT` and set the build number to "BETA" ill the new version by `1.0.BETA`? And then if I do another release with no build number, is the verison `1.0.0`? Is the next release after that `1.0.1` or is that `1.0.BETA` version having an influence?

I must say though, as a user of `mvn versions:display-dependency-updates` I find it really annoying when people don't use numeric version numbers.

**Comment by marcelstoer** [comment]
If I got your last comment right then I guess the "lack of understanding" on your behalf is a (maybe) simple misunderstanding. This change here only affects the build number i.e. anything you specify _after_ the - ideally semantic - version. This is limited to the part that the plugin modifies by automatically incrementing. So, for your example it wouldn't be `1.0.BETA` but rather `1.0.0-BETA`.

**Comment by danielflower** [comment]
Yeah, I guess I never understood this PR then.

So if I have `1.0-SNAPSHOT` and set build number to `BETA` it becomes `1.0.0-BETA`. The next release (with no custom build number) is `1.0.1` etc?

So this is more of a suffix to the version than a build number? Then why not leave it as a `long` and have a new property: `versionSuffix` or some-such?
